junebug/amqp.py

```diff
@@ -0,0 +1,153 @@
+from vumi.utils import vumi_resource_path
+from vumi.service import get_spec
+
+from junebug.error import JunebugError
+
+
+class AmqpConnectionError(JunebugError):
+    '''Exception that is raised whenever a message is attempted to be send,
+    but no amqp connection is available to send it on'''
+    name = 'AmqpConnectionError'
+    description = 'amqp connection error'
+    code = http.INTERNAL_SERVER_ERROR
+
+
+class MessageSender(MultiService):
+    '''Keeps track of the amqp connection and can send messages. Raises an
+    exception if a message is sent when there is no amqp connection'''
+    def __init__(self, specfile, amqp_config):
+        super(MessageSender, self).__init__()
+        self.amqp_config = amqp_config
+        self.factory = AmqpFactory(
+            specfile, amqp_config, self._connected_callback,
+            self._disconnected_callback)
+
+    def startService(self):
+        super(MessageSender, self).startService()
+        self.amqp_service = TCPClient(
+            self.amqp_config['hostname'], self.amqp_config['port'],
+            self.factory)
+        self.amqp_service.setServiceParent(self)
+
+    def _connected_callback(self, client):
+        self.client = client
+
+    def _disconnected_callback(self):
+        self.client = None
+
+    def send_message(self, message, **kwargs):
+        if not hasattr(self, 'client') or self.client is None:
+            raise AmqpConnectionError(
+                'Message not sent, AMQP connection error.')
+        return self.client.publish_message(message, **kwargs)
+
+
+class AmqpFactory(ReconnectingClientFactory, object):
+    def __init__(
+            self, specfile, amqp_config, connected_callback,
+            disconnected_callback):
+        '''Factory that creates JunebugAMQClients.
+        specfile - string of specfile name
+        amqp_config - connection details for amqp server
+        '''
+        self.connected_callback, self.disconnected_callback = (
+            connected_callback, disconnected_callback)
+        self.amqp_config = amqp_config
+        self.spec = get_spec(vumi_resource_path(specfile))
+        self.delegate = TwistedDelegate()
+
+    def buildProtocol(self, addr):
+        amqp_client = JunebugAMQClient(
+            self.delegate, self.amqp_config['vhost'],
+            self.spec, self.amqp_config.get('heartbeat', 0))
+        amqp_client.factory = self
+        self.resetDelay()
+        return amqp_client
+
+    def clientConnectionFailed(self, connector, reason):
+        log.err("AmqpFactory connection failed (%s)" % (
+            reason.getErrorMessage(),))
+        super(AmqpFactory, self).clientConnectionFailed(connector, reason)
+
+    def clientConnectionLost(self, connector, reason):
+        log.err("AmqpFactory client connection lost (%s)" % (
+            reason.getErrorMessage(),))
+        self.disconnected_callback()
+        super(AmqpFactory, self).clientConnectionLost(connector, reason)
```

Comment:
Turns out it's quite difficult. Tried using a `_FakeConnector`, but the `clientConnectionFailed` of the `ReconnectingClientFactory` does a `clock.callLater(...)`, which makes it quite difficult to test. Ideas on testing this?